src/plugins/shared_dictionaries_plugin/shared_dictionaries_column_processor.cpp

```diff
@@ -0,0 +1,291 @@
+        merge_plan->shared_dictionary = best_shared_dictionary;
+        _add_segment_chunk_pair(*merge_plan, current_segment_chunk_pair, false);
+        merged_current_dictionary = true;
+      } else if (previous_segment_chunk_pair_opt) {
+        // Check with previous segment
+        const auto shared_dictionary_with_previous =
+            _compare_with_previous_dictionary(current_dictionary, *previous_segment_chunk_pair_opt, allocator);
+        if (shared_dictionary_with_previous) {
+          // Merge with previous dictionary
+          const auto new_merge_plan = std::make_shared<MergePlan>(shared_dictionary_with_previous);
+          _add_segment_chunk_pair(*new_merge_plan, current_segment_chunk_pair, false);
+          _add_segment_chunk_pair(*new_merge_plan, *previous_segment_chunk_pair_opt, false);
+          merge_plans.emplace_back(new_merge_plan);
+          merged_current_dictionary = true;
+        }
+      }
+
+      // Save unmerged current dictionary for possible later merge
+      previous_segment_chunk_pair_opt =
+          merged_current_dictionary ? std::nullopt : std::make_optional(current_segment_chunk_pair);
+    }
+  }
+  _process_merge_plans(merge_plans, allocator);
+}
+
+template <typename T>
+void SharedDictionariesColumnProcessor<T>::_initialize_merge_plans(
+    std::vector<std::shared_ptr<MergePlan>>& merge_plans) {
+  auto shared_dictionaries_map = std::map<std::shared_ptr<const pmr_vector<T>>, std::shared_ptr<MergePlan>>{};
+  const auto chunk_count = table->chunk_count();
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    const auto chunk = table->get_chunk(chunk_id);
+    const auto segment = chunk->get_segment(column_id);
+
+    const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(segment);
+    if (dictionary_segment && dictionary_segment->uses_dictionary_sharing()) {
+      stats->num_existing_merged_dictionaries++;
+      const auto shared_dictionary = dictionary_segment->dictionary();
+      const auto [iter, inserted] =
+          shared_dictionaries_map.try_emplace(shared_dictionary, std::make_shared<MergePlan>(shared_dictionary));
+      const auto segment_chunk_pair = std::make_pair(dictionary_segment, chunk);
+      const auto merge_plan = iter->second;
+      _add_segment_chunk_pair(*merge_plan, segment_chunk_pair, true);
+    }
+  }
+
+  stats->num_existing_shared_dictionaries = static_cast<uint32_t>(shared_dictionaries_map.size());
+
+  for (const auto& [key, merge_plan] : shared_dictionaries_map) {
+    merge_plans.emplace_back(merge_plan);
+  }
+}
+
+template <typename T>
+std::pair<int32_t, std::shared_ptr<const pmr_vector<T>>>
+SharedDictionariesColumnProcessor<T>::_compare_with_existing_merge_plans(
+    const std::shared_ptr<const pmr_vector<T>> current_dictionary,
+    const std::vector<std::shared_ptr<MergePlan>>& merge_plans, const PolymorphicAllocator<T>& allocator) {
+  auto best_merge_plan_index = -1;
```

Comment:
Could we communicate the intent clearer by using and optional here instead of passing `-1` to encode no best merge plan?

Reply:
Fixed in https://github.com/hyrise/hyrise/pull/2394/commits/7f88317d0867f9bfd92688f5061daea73a39be35